### DISCUSSION TITLE
Nodriver proxy with authentication

### DISCUSSION THREAD

**Discussion by strukiii** [ultrafunkamsterdam/undetected-chromedriver#1798] [Q&A] [answered]
Hi all, I'm putting this here since there is no nodriver discussions section.  I'm trying to get the browser setup with an authenticated proxy but chrome does not support credentials via browser args.  I found [this,](https://superuser.com/questions/902620/google-chrome-proxy-settings-with-username-and-password) where the answer from Arkadiy Bolotov is interesting.  There are the same cdp methods as in his post available in [Network](https://github.com/ultrafunkamsterdam/nodriver/blob/main/nodriver/cdp/network.py#L3280) though they are deprecated for those in [Fetch](https://github.com/ultrafunkamsterdam/nodriver/blob/main/nodriver/cdp/fetch.py#L215).  Here is my current attempt I put together this afternoon, which does intercept the requests however no authRequired events are fired as the documentation suggests.  

```
class Scraper:
    main_tab: tab

    def __init__(self):
        loop().run_until_complete(self.run())

    async def run(self):
        browser = await start(
            browser_args=[f"--proxy-server={PROXY}"],
        )
        self.main_tab = browser.main_tab
        await self.main_tab.send(cdp.fetch.enable(handle_auth_requests=True))
        self.main_tab.add_handler(cdp.fetch.RequestPaused, self.req_paused)
        self.main_tab.add_handler(cdp.fetch.AuthRequired, self.auth_challenge_handler)
        page = await browser.get('https://www.whatismyip.com/')
        await asyncio.sleep(6000)


    async def auth_challenge_handler(self, event: cdp.fetch.AuthRequired):
        # Split the credentials
        # Respond to the authentication challenge
        await self.main_tab.send(cdp.fetch.continue_with_auth(
            request_id=event.request_id,
            auth_challenge_response=cdp.fetch.AuthChallengeResponse(
                response="ProvideCredentials",
                username=USERNAME,
                password=PASSWORD
            )
        ))


    async def req_paused(self, event: cdp.fetch.RequestPaused):
        await self.main_tab.send(cdp.fetch.continue_request(request_id=event.request_id))


if __name__ == '__main__':
    Scraper()
```

**Comment by WindBora** [comment, marked as answer]
Hi

Here 2 problems arises:
1. You add handlers after enabling domain (internal implementation overrides enabling domain) - that's the reason why handlers do not receive an event
2. When you will finally receive an event, your response will block all processes

Solutions:
1. Add handlers before enabling domain. Only after adding them do enable domain
4. For such blocking requests there is a must to just use `fire-and-forget` method e.g. do not await for sending response on `req_paused` or `auth_required`. Just wrap the call into `asyncio.create_task`

Example of working code based of of yours:
```python
class Scraper:
    main_tab: uc.Tab

    def __init__(self):
        uc.loop().run_until_complete(self.run())

    async def run(self):
        browser = await uc.start(
            browser_args=[f"--proxy-server={PROXY}"],
        )
        self.main_tab = await browser.get("draft:,")
        self.main_tab.add_handler(uc.cdp.fetch.RequestPaused, self.req_paused)
        self.main_tab.add_handler(
            uc.cdp.fetch.AuthRequired, self.auth_challenge_handler
        )
        await self.main_tab.send(uc.cdp.fetch.enable(handle_auth_requests=True))
        page = await browser.get("https://www.whatismyip.com/")
        await asyncio.sleep(6000)

    async def auth_challenge_handler(self, event: uc.cdp.fetch.AuthRequired):
        # Split the credentials
        # Respond to the authentication challenge
        asyncio.create_task(
            self.main_tab.send(
                uc.cdp.fetch.continue_with_auth(
                    request_id=event.request_id,
                    auth_challenge_response=uc.cdp.fetch.AuthChallengeResponse(
                        response="ProvideCredentials",
                        username=USERNAME,
                        password=PASSWORD,
                    ),
                )
            )
        )

    async def req_paused(self, event: uc.cdp.fetch.RequestPaused):
        asyncio.create_task(
            self.main_tab.send(
                uc.cdp.fetch.continue_request(request_id=event.request_id)
            )
        )


if __name__ == "__main__":
    Scraper()
```

**Reply by strukiii** [reply]
Great work!  Just tried it out, works perfectly so far.

**Reply by flaming-chameleon** [reply]
How can i change headers in this way for request?

**Reply by rex1de** [reply]
not working for me

**Reply by cu0302pid** [reply]
for me also doesnt work

**Reply by mushifali** [reply]
Thanks a lot! It works perfectly!

**Reply by 78z0** [reply]
doesnt work for me, what is the "PROXY" variable format?

**Reply by mushifali** [reply]
@78z0 It's `{ip_address}:{port}`

**Reply by tidal10** [reply]
Would you mind providing your imports?

**Reply by mushifali** [reply]
```py
import asyncio
import nodriver as uc
```

These should be enough.

**Reply by tidal10** [reply]
Do you have a way to run multiple proxy sessions on a threaded environment with different proxies?

**Reply by epicwhale** [reply]
This works for me with non-ssl domains/pages, but fails when it's an https:// url I am trying to access. I get the following (screenshot below)... 

Wondering is there some flag or extra args you needed to set somewhere to workaround this by any chance?  @mushifali @strukiii @WindBora 

![image](https://github.com/user-attachments/assets/b8570d4f-d05b-4510-8a5c-891dc9d27903)

**Reply by mushifali** [reply]
@epicwhale It is also working for https:// domains for me. Can you please share your code?

**Reply by epicwhale** [reply]
@mushifali looks like it was an issue with the proxy I was using, thanks for confirming! A better proxy seemed to do it for me too.

**Reply by nathanfallet** [reply]
I cannot get this example working. Actually this line made the first page loading but any subsequent calls paused:
```py
await self.main_tab.send(uc.cdp.fetch.enable(handle_auth_requests=True))
```
For example, if you go to a website with html/css/js, the main html will load, but all css/js assets will be marked as pending and never continues. This is due to `req_paused` being called for the html page but never called again for all those assets (so we cannot continue the request)
Do you have any way to solve this? Maybe someone else already encountered this issue.

**Comment by SSujitX** [comment]
```
import nodriver as uc
from nodriver.cdp import fetch

PROXYHOST = "aaaaaaaaaaaaaa.ika.na.pyproxy.io"
PROXYPORT = "2000"
PROXYUSERNAME = "aaa"
PROXYPASSWORD = "bbbb"
PROXY = f"http://{PROXYHOST}:{PROXYPORT}"

async def setup_proxy(username, password, tab):
    async def auth_challenge_handler(event: fetch.AuthRequired):
        # Respond to the authentication challenge
        await tab.send(
            fetch.continue_with_auth(
                request_id=event.request_id,
                auth_challenge_response=fetch.AuthChallengeResponse(
                    response="ProvideCredentials",
                    username=username,
                    password=password,
                ),
            )
        )

    async def req_paused(event: fetch.RequestPaused):
        # Continue with the request
        await tab.send(fetch.continue_request(request_id=event.request_id))

    # Add handlers for fetch events
    tab.add_handler(
        fetch.RequestPaused, lambda event: asyncio.create_task(req_paused(event))
    )
    tab.add_handler(
        fetch.AuthRequired,
        lambda event: asyncio.create_task(auth_challenge_handler(event)),
    )

    # Enable fetch domain with auth requests handling
    await tab.send(fetch.enable(handle_auth_requests=True))


    # -------- PROXY -------- #
    browser = await uc.start(browser_args=[f"--proxy-server={PROXY}"])
    main_tab = await browser.get("draft:,")
    await setup_proxy(PROXYUSERNAME, PROXYPASSWORD, main_tab)
    ipPage = await browser.get("https://www.myexternalip.com/raw")
    await asyncio.sleep(2)
    ip = await ipPage.evaluate("document.body.textContent.trim()")
    print(f">= IP Address: {ip}")
    # -------- PROXY -------- #


```

**Reply by aoddspotato** [reply]
This is awesome

**Reply by kashif-ghafoor** [reply]
Thanks Brother. Saved a lot of my time.

**Reply by C0uchP0tat0** [reply]
Thanks, bro! This is exactly what I needed. This solution works for Zendriver too.

**Reply by nickloganfarmer** [reply]
Thanks for the help. It was driving me crazy why my regularly formatted proxy address string wasn't working.

**Comment by LuisGMM** [comment]
Hi all, I have been trying to use a proxy with nodriver for a couple of days now, without success. I tried the suggested solutions but I keep getting this error in the chrome tab: `ERR_NO_SUPPORTED_PROXIES`

Image:
![image](https://github.com/user-attachments/assets/372a5110-61fa-4920-9b0a-79dfbb0d40aa)


In my case, in am trying to use smartproxy residential proxies. A snippet of working code with `requests` would be this:
```python
import requests
url = 'https://ip.smartproxy.com/json'
username = 'some_user'
password = 'some_password'
proxy = f"http://{username}:{password}@es.smartproxy.com:10000"
result = requests.get(url, proxies = {
    'http': proxy,
    'https': proxy
})
print(result.text)
```


A working example with `undetected_chromedriver` & `selenium_wire` would be this:
```python
import seleniumwire.undetected_chromedriver as uc

# Proxy server details
PROXYUSERNAME = 'some_user'
PROXYPASSWORD = 'some_password'
PROXYPORT = '10000'

# define your proxy credentials
proxy_username = PROXYUSERNAME
proxy_password = PROXYPASSWORD
proxy_host = 'es.smartproxy.com'
proxy_port = PROXYPORT

# form the proxy address
proxy_address = f'http://{proxy_username}:{proxy_password}@{proxy_host}:{proxy_port}'

# add the proxy address to proxy options
proxy_options = {
    'proxy': {
        'http': proxy_address,
        'https': proxy_address,
    }
}

if __name__ == '__main__':
    # set Chrome options
    options = uc.ChromeOptions()

    # run Chrome in headless mode
    options.headless = True
    options.add_argument('--ignore-certificate-errors')

    # create a Chrome instance with the proxy options
    driver = uc.Chrome(
        seleniumwire_options=proxy_options,
        options=options,
        use_subprocess=False,
    )

    # visit the test URL to check your proxy IP
    driver.get('https://httpbin.io/ip')
    ip_address = driver.find_element(By.TAG_NAME, 'body').text
    print(ip_address)

    # close the browser
    driver.quit()
```


However, when using these credentials with `nodriver` I get the previous error. Any ideas anyone?

Ps: @strukiii @WindBora @epicwhale @SSujitX

**Reply by SSujitX** [reply]
> Hi all, I have been trying to use a proxy with nodriver for a couple of days now, without success. I tried the suggested solutions but I keep getting this error in the chrome tab: `ERR_NO_SUPPORTED_PROXIES`

> However, when using these credentials with `nodriver` I get the previous error. Any ideas anyone?
> 
> Ps: @strukiii @WindBora @epicwhale @SSujitX

First, tell me which library you want to use: nodriver, seleniumwire, seleniumbase, or requests?

**Reply by LuisGMM** [reply]
Sure, maybe I explained myself wrong, sorry about that :)
I would love to use **nodriver with a proxy**, as described above by other colleagues, however I cannot manage to make it work as I could using `requests` or `undetected_chromedriver` with `seleniumwire`

**Reply by ultrafunkamsterdam** [reply]
`await start(browser_args=["--proxy-server=socks5://127.0.0.1:9150"])`

**Reply by LuisGMM** [reply]
What does that mean @ultrafunkamsterdam ?

**Reply by felipemedlev** [reply]
Any luck?

**Reply by ultrafunkamsterdam** [reply]
> What does that mean @ultrafunkamsterdam ?

I would love to use nodriver with a proxy,

**Comment by BarisSenel** [comment]
Hi, here's what I usually do for authentication-required SOCKS5 proxies with Chromium. I use [GOST](https://github.com/go-gost/gost)
It can be used on linux/windows
```python
import subprocess

def start_gost_proxy(proxy, rport):
    ip, port, user, password = proxy.split(":")
    gost_proxy = f"socks5://{user}:{password}@{ip}:{port}"
    gost_command = ["gost", "-L", f":{rport}", "-F", gost_proxy]
    subprocess.Popen(gost_command, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    print(f"Started GOST proxy with {gost_proxy} on port {rport}")

async def login(email, password):
    start_gost_proxy("YourProxyIP:YourProxyPort:Username:Password", 2000)
    browser = await start(headless=False, lang="en-US", browser_args=["--proxy-server=socks5://localhost:2000"])
```

**Reply by sufyanhamid** [reply]
@BarisSenel will it work with the http datacenter proxies?

**Reply by BarisSenel** [reply]
Yes it would you can look up to GOST documents it supports both socks5 / http https://gost.run/

**Reply by diyorjon-olimjonov** [reply]
it worked, thanks

**Comment by ultrafunkamsterdam** [comment]
Proxy with auth solved in latest version, you can also use per window/tab another proxy

[see here](https://github.com/ultrafunkamsterdam/nodriver/issues/4#issuecomment-2755090686)

**Reply by danofun** [reply]
Thank you for implementing this feature! Unfortunately, I'm receiving an error.

```
import nodriver
import time

async def main():

    browser = await nodriver.start(
        headless=False,
    )
    page = await browser.create_context('https://www.ipchicken.com', proxy_server='http://user:pass@ip:port')
    time.sleep(15)
    

if __name__ == '__main__':
    # since asyncio.run never worked (for me)
    nodriver.loop().run_until_complete(main())
```


is producing the following error on both Debian 12 and macOS 15.3.2 using either python 3.11 or 3.13.

```
Traceback (most recent call last):
  File "proxy_test.py", line 16, in <module>
    nodriver.loop().run_until_complete(main())
  File "/usr/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "proxy_test.py", line 10, in main
    page = await browser.create_context('https://www.ipchicken.com', proxy_server='http://user:pass@ip:port')
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/nodriver/core/browser.py", line 307, in create_context
    fw = util.ProxyForwarder(proxy_server=proxy_server)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/nodriver/core/util.py", line 4589, in __init__
    self.port = util.free_port()
                ^^^^
NameError: name 'util' is not defined. Did you mean: 'shutil'?
successfully removed temp profile /tmp/uc_ail1vv85
```

Is this implementation for socks5 authenticated proxies only?

**Reply by BarisSenel** [reply]
@ultrafunkamsterdam  @danofun 
Enter the util.py from "C:\Users\{Username}\AppData\Local\Programs\Python\Python313\Lib\site-packages\nodriver\core"
inside util.py there will be this line
```self.port = util.free_port()```
change it to 
```self.port = free_port()```
fixed for me

**Reply by ultrafunkamsterdam** [reply]
Damn. Yes I moved the custom proxy to util file laat minute . Thanks will fix in repo in 60 minutes

**Reply by ultrafunkamsterdam** [reply]
> Thank you for implementing this feature! Unfortunately, I'm receiving an error.
> 
> ```
> import nodriver
> import time
> 
> async def main():
> 
>     browser = await nodriver.start(
>         headless=False,
>     )
>     page = await browser.create_context('https://www.ipchicken.com', proxy_server='http://user:pass@ip:port')
>     time.sleep(15)
>     
> 
> if __name__ == '__main__':
>     # since asyncio.run never worked (for me)
>     nodriver.loop().run_until_complete(main())
> ```
> 
> is producing the following error on both Debian 12 and macOS 15.3.2 using either python 3.11 or 3.13.
> 
> ```
> Traceback (most recent call last):
>   File "proxy_test.py", line 16, in <module>
>     nodriver.loop().run_until_complete(main())
>   File "/usr/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
>     return future.result()
>            ^^^^^^^^^^^^^^^
>   File "proxy_test.py", line 10, in main
>     page = await browser.create_context('https://www.ipchicken.com', proxy_server='http://user:pass@ip:port')
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/dist-packages/nodriver/core/browser.py", line 307, in create_context
>     fw = util.ProxyForwarder(proxy_server=proxy_server)
>          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/dist-packages/nodriver/core/util.py", line 4589, in __init__
>     self.port = util.free_port()
>                 ^^^^
> NameError: name 'util' is not defined. Did you mean: 'shutil'?
> successfully removed temp profile /tmp/uc_ail1vv85
> ```
> 
> Is this implementation for socks5 authenticated proxies only?

Yes socks5 only atm

**Reply by zerxdapro** [reply]
> Proxy with auth solved in latest version, you can also use per window/tab another proxy
> 
> [see here](https://github.com/ultrafunkamsterdam/nodriver/issues/4#issuecomment-2755090686)

'Browser' object has no attribute 'create_context'

**Comment by Pavlo165** [comment]
I'm trying to use a SOCKS5 proxy with authentication in `nodriver`. However, when I set the proxy in `create_context`, I get an error and the browser fails to start the context.

# Code

```python
async def init_browser():
    browser = await uc.start(
        headless=False,
        browser_args=[
            "--start-maximized"
        ]
    )
    proxied_tab = await browser.create_context(
        proxy_server = "socks5://user:password@dc.oxylabs.io:8005",
        proxy_bypass_list = ["localhost"]
    )
    await proxied_tab.get('https://whatismyip.com')
    return browser, proxied_tab

async def main():
    browser, main_tab = await init_browser()
    await browser.wait(3600)

if __name__ == "__main__":
    uc.loop().run_until_complete(main())
    
```

# Error
```
2025-06-21 23:47:18,903 INFO [2848] nodriver.core.browser [browser.py:440] - enabling autodiscover targets
2025-06-21 23:47:18,917 INFO [2848] nodriver.core.util [util.py:4599] - socks proxy with authentication is requested : socks5://user:password@dc.oxylabs.io:8005
2025-06-21 23:47:18,921 INFO [2848] nodriver.core.util [util.py:4602] - starting forward proxy on 127.0.0.1:64927
2025-06-21 23:47:18,934 INFO [2848] nodriver.core.util [util.py:4603] - which forwards to socks5://user:password@dc.oxylabs.io:8005
Traceback (most recent call last):
  File "C:\Users\mothu\OneDrive\Рабочий стол\nodriver\parser\parser.py", line 45, in <module>
    uc.loop().run_until_complete(main())
  File "C:\Users\mothu\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "C:\Users\mothu\OneDrive\Рабочий стол\nodriver\parser\parser.py", line 38, in main
    browser, main_tab = await init_browser()
                        ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mothu\OneDrive\Рабочий стол\nodriver\parser\parser.py", line 19, in init_browser
    proxied_tab = await browser.create_context(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mothu\AppData\Local\Programs\Python\Python311\Lib\site-packages\nodriver\core\browser.py", line 321, in create_context
    ctx: cdp.browser.BrowserContextID = await self.connection.send(
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\mothu\AppData\Local\Programs\Python\Python311\Lib\site-packages\nodriver\core\connection.py", line 518, in send
    return await tx
           ^^^^^^^^
nodriver.core.connection.ProtocolException: Invalid parameters [code: -32602]
2025-06-21 23:47:18,989 INFO [2848] nodriver.core.browser [browser.py:683] - terminated browser with pid 12416 successfully
successfully removed temp profile C:\Users\mothu\AppData\Local\Temp\uc_y477s70t
```
Does anyone know how to properly set a SOCKS5 proxy with authentication in nodriver? Any help or advice is appreciated.

**Reply by erik9631** [reply]
Same here. Is this issue tracked?

**Reply by yashjai2005-bs** [reply]
wondering if anyone found a fix for it? @ultrafunkamsterdam

I also tried using proxy_server browser arg https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/2185 but that doesn't work either...

**Comment by nathanfallet** [comment]
We cannot get a proxy with auth working directly in the browser, because auth is not supported in the URL and you need to pause a request to handle it. The issue we got was that every other requests were paused when we didn't want them paused, so we could not do anything.

The way we made it work was by using [gost](https://github.com/go-gost/gost) to start a local proxy that handles auth for us, and then connecting the browser to the local proxy without auth:

```python
# Somewhere before starting the browser
# In your case, url = "socks5://user:password@dc.oxylabs.io:8005"
gost_command = ["gost", "-L", f":{port}", "-F", url]
subprocess.Popen(gost_command, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

# When starting the browser
browser_args = [
    f"--proxy-server=socks5://localhost:{port}",
    # ...
]
```

**Reply by juhacz** [reply]
Similar repository like nodriver or zendriver https://github.com/kaliiiiiiiiii/Selenium-Driverless The program advises the proxy with authorization perfectly. Pay attention to the set_proxy() method.

**Comment by ultrafunkamsterdam** [comment]
Just don't use url in create_context. Just call it. Get the tab instance and call get(url) on it.

**Reply by bourmaofficial** [reply]
Hi @ultrafunkamsterdam, it works fine, but crashes if the site redirects to an IPv6-only host (e.g. apiv6.whatismyip.com). The error looks like this:

`ip_addr = socket.gethostbyname(hostname)
socket.gaierror: [Errno 11001] getaddrinfo failed
`

This happens because socket.gethostbyname() only supports IPv4. Switching to socket.getaddrinfo() would handle both IPv4 and IPv6 and prevent the crash.
